exiftran.man
.TH EXIFTRAN 1 "(c) 2002-2012 Gerd Hoffmann" "EXIFTRAN 2.09" "Transform digital camera jpeg images"
\#
\#
.SH NAME
exiftran - transform digital camera jpeg images
\#
\#
.SH SYNOPSIS
.TP
\fBexiftran\fP [\fIoptions\fP] \fIfile1 file2 ... fileN\fP
.TP
\fBexiftran\fP -i [\fItransform options\fP] [\fIothers options\fP] \fIfile1 file2 ... fileN\fP
.TP
\fBexiftran\fP -o \fIoutputfile\fP [\fItransform options\fP] [\fIother options\fP] \fIinputfile\fP
.TP
\fBexiftran\fP -d \fIfile1 file2 ... fileN\fP > \fIexifinfo\fP
\#
\#
.SH DESCRIPTION
.BR Exiftran
is a command line utility to transform digital camera \fIjpeg\fP images. It can do lossless
rotations like
.BR "jpegtran" "(1),"
but unlike
.BR "jpegtran" "(1)"
it cares about the \fIEXIF\fP data: It can rotate images automatically by
checking the exif orientation tag; it updates the exif informations if needed
(image dimension, orientation); it also rotates the exif thumbnail. It can
process multiple images at once.
\#
\#
.SH "TRANSFORM OPTIONS"
.TP
.B -a
Automatic (using exif orientation tag).
.TP
.B -9
Rotate by 90 degrees clockwise.
.TP
.B -1
Rotate by 180 degrees clockwise.
.TP
.B -2
Rotate by 270 degrees clockwise.
.TP
.B -f
Mirror image vertically (top / bottom).
.TP
.B -F
Mirror image horizontally (left to right).
.TP
.B -t
Transpose (across UL-to-LR corner).
.TP
.B -T
Transverse (across UR-to-LL corner).
.TP
.B -nt
Don't transform exif thumbnail.
.TP
.B -ni
Don't transform jpeg image. You might need this or the \fB-nt\fP option to fixup
things in case you transformed the image with some utility which ignores the
exif thumbnail. Just generating a new thumbnail with \fB-g\fP is another way to
fix it.
.TP
.B -no
Don't update the orientation tag. By default
.BR exiftran
sets the orientation to "1" (no transformation needed) to avoid other exif-aware
applications try to rotate the already-rotated image again.
.TP
.B -np
Don't pare lost edges. By default
.BR exiftran
don't preserve image size of the images that do not meet a
multiple of 8 pixels. He prefers to cut a strip of a few pixels rather than
offering a damaged image. Use this option if you want them all the same.
\#
\#
.SH "OTHER OPTIONS"
.TP
.B -h
Print a short help text.
.TP
.B -d
Dump exif data for the file(s).
.TP
.BI "-c" "\ text"
Set jpeg comment tag to \fItext\fP.
.TP
.B -g
(re)generate exif thumbnail.
.TP
.BI "-o" "\ file"
Specify output \fIfile\fP. Only one input file is allowed in this mode.
.TP
.B -i
Enable inplace editing of the images.
.BR Exiftran
allows multiple input files then. You must specify either this option or a
output file with \fB-o\fP for all operations which modify the image (i.e.
everything but \fB-d\fP right now).
.TP
.B -b
Create a backup file when doing inplace editing (imply \fB-i\fP).
.TP
.B -p
Preserve timestamps (atime + mtime) when doing inplace editing (imply \fB-i\fP).
\#
\#
.SH EXAMPLES
Autorotate all jpeg files in the current directory:
.P
.in +4n
   \fIexiftran\ -ai\ *.jpeg\fP
.in
\#
\#
.SH "SEE ALSO"
.BR exif (1),
.BR exiftags (1),
.BR jpegtran (1)
\#
\#
.SH AUTHOR
Gerd Hoffmann
.BR <gerd@kraxel.org>
\#
\#
.SH COPYRIGHT
Copyright (c) 2002-2012 Gerd Hoffmann <gerd@kraxel.org>
.P
This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.
.P
This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.
.P
You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 675 Mass Ave,
Cambridge, MA 02139, USA.


exiftran.c
/*
 * Transform digital camera jpeg images
 *
 * (c) 2002-2012 Gerd Hoffmann <gerd@kraxel.org>
 */

#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>

#include <libexif/exif-data.h>

#include <jpeglib.h>
#include "transupp.h"		/* Support routines for jpegtran */
#include "jpegtools.h"
#include "genthumbnail.h"

/* ---------------------------------------------------------------------- */

static void dump_exif(FILE *out, ExifData *ed)
{
    const char *title, *value;
#ifdef HAVE_NEW_EXIF
    char buffer[256];
#endif
    ExifEntry  *ee;
    int tag,i;

    for (i = 0; i < EXIF_IFD_COUNT; i++) {
	fprintf(out,"   ifd %s\n", exif_ifd_get_name (i));
	for (tag = 0; tag < 0xffff; tag++) {
	    title = exif_tag_get_title(tag);
	    if (!title)
		continue;
	    ee = exif_content_get_entry (ed->ifd[i], tag);
	    if (NULL == ee)
		continue;
#ifdef HAVE_NEW_EXIF
	    value = exif_entry_get_value(ee, buffer, sizeof(buffer));
#else
	    value = exif_entry_get_value(ee);
#endif
	    fprintf(out,"      0x%04x  %-30s %s\n", tag, title, value);
	}
    }
    if (ed->data && ed->size)
	fprintf(out,"   thumbnail\n      %d bytes data\n", ed->size);
}

static int dump_file(FILE *out, char *filename)
{
    ExifData   *ed;

    ed = exif_data_new_from_file (filename);
    if (NULL == ed) {
	fprintf(stderr,"%s: no EXIF data\n",filename);
	return -1;
    }

    fprintf(out,"%s\n",filename);
    dump_exif(out,ed);
    fprintf(out,"--\n");

    exif_data_unref (ed);
    return 0;
}

/* ---------------------------------------------------------------------- */

#define THUMB_MAX 65536

static void
usage(FILE *fp, char *name)
{
    char *h;

    if (NULL != (h = strrchr(name, '/')))
	name = h+1;
    fprintf(fp,
	    "usage: %s [ options ] file\n"
	    "\n"
	    "transform options:\n"
	    "  -a         automatic (using exif orientation tag)\n"
	    "  -9         rotate by 90 degrees clockwise\n"
	    "  -1         rotate by 180 degrees clockwise\n"
	    "  -2         rotate by 270 degrees clockwise\n"
	    "  -f         mirror image vertically (top / bottom)\n"
	    "  -F         mirror image horizontally (left to right)\n"
	    "  -t         transpose (across UL-to-LR corner)\n"
	    "  -T         transverse (across UR-to-LL corner)\n"
	    "\n"
	    "  -nt        don't rotate exif thumbnail\n"
	    "  -ni        don't rotate jpeg image\n"
	    "  -no        don't update the orientation tag\n"
	    "  -np        don't pare lost edges\n"
	    "\n"
	    "other options:\n"
	    "  -h         print this help text\n"
	    "  -d         dump exif data\n"
	    "  -c <text>  create/update comment\n"
	    "  -g         (re)generate thumbnail\n"
	    "  -o <file>  output file\n"
	    "  -i         change files inplace\n"
	    "  -b         create a backup file (with -i)\n"
	    "  -p         preserve timestamps  (with -i)\n"
	    "\n"
	    "-- \n"
	    "(c) 2002-2012 Gerd Hoffmann <gerd@kraxel.org> [SUSE Labs]\n",
	    name);
}

int main(int argc, char *argv[])
{
    JXFORM_CODE transform = JXFORM_NONE;
    unsigned char *comment = NULL;
    unsigned char *outfile = NULL;
    unsigned char *thumbnail = NULL;
    int tsize = 0;
    int inplace = 0;
    unsigned int flags =
	JFLAG_TRANSFORM_IMAGE     |
	JFLAG_TRANSFORM_THUMBNAIL |
	JFLAG_TRANSFORM_TRIM      |
	JFLAG_UPDATE_ORIENTATION;
    int dump = 0;
    int i, c, rc;

    for (;;) {
	c = getopt(argc, argv, "hbpid912fFtTagc:o:n:");
	if (c == -1)
	    break;
	switch (c) {
	case '9':
	    transform = JXFORM_ROT_90;
	    break;
	case '1':
	    transform = JXFORM_ROT_180;
	    break;
	case '2':
	    transform = JXFORM_ROT_270;
	    break;
	case 'f':
	    transform = JXFORM_FLIP_V;
	    break;
	case 'F':
	    transform = JXFORM_FLIP_H;
	    break;
	case 't':
	    transform = JXFORM_TRANSPOSE;
	    break;
	case 'T':
	    transform = JXFORM_TRANSVERSE;
	    break;
	case 'a':
	    transform = -1; /* automagic */
	    break;

	case 'n':
	    /* don't ... */
	    switch (optarg[0]) {
	    case 't':
		flags &= ~JFLAG_TRANSFORM_THUMBNAIL;
		break;
	    case 'i':
		flags &= ~JFLAG_TRANSFORM_IMAGE;
		break;
	    case 'o':
		flags &= ~JFLAG_UPDATE_ORIENTATION;
		break;
	    case 'p':
		flags &= ~JFLAG_TRANSFORM_TRIM;
		break;
	    default:
		fprintf(stderr,"unknown option -n%c\n",optarg[0]);
		exit(1);
	    }
	    break;

	case 'c':
	    flags |= JFLAG_UPDATE_COMMENT;
	    comment = optarg;
	    break;
	case 'g':
	    flags |= JFLAG_UPDATE_THUMBNAIL;
	    break;
	case 'o':
	    outfile = optarg;
	    break;
	case 'd':
	    dump = 1;
	    break;

	case 'b':
	    flags |= JFLAG_FILE_BACKUP;
	    break;
	case 'p':
	    flags |= JFLAG_FILE_KEEP_TIME;
	    break;
	case 'i':
	    inplace = 1;
	    break;

	case 'h':
	    usage(stdout,argv[0]);
	    exit(0);
	default:
	    usage(stderr,argv[0]);
	    exit(1);
	}
    }

    /* sanity checks on the arguments */
    if (optind == argc) {
	fprintf(stderr,
		"no image file specified (try -h for more info)\n");
	exit(1);
    }

    /* read-only stuff */
    if (dump) {
	rc = 0;
	for (i = optind; i < argc; i++) {
	    if (0 != dump_file(stdout,argv[i]))
		rc = 1;
	}
	return rc;
    }

    /* r/w sanity checks */
    if (NULL != outfile && optind+1 > argc) {
	fprintf(stderr,
		"when specifying a output file you can process\n"
		"one file at a time only (try -h for more info).\n");
	exit(1);
    }
    if (NULL == outfile && 0 == inplace) {
	fprintf(stderr,
		"you have to either specify a output file (-o <file>)\n"
		"or enable in-place editing (-i). Try -h for more info.\n");
	exit(1);
    }
    if (JXFORM_NONE == transform &&
	!(flags & JFLAG_UPDATE_COMMENT) &&
	!(flags & JFLAG_UPDATE_THUMBNAIL)) {
	fprintf(stderr,
		"What do you want to do today?  Neither a new comment nor a\n"
		"transformation operation was specified (try -h for more info).\n");
	exit(1);
    }

    /* do actual update work */
    if (outfile) {
	if (flags & JFLAG_UPDATE_THUMBNAIL) {
	    thumbnail = malloc(THUMB_MAX);
	    tsize = create_thumbnail(argv[optind],thumbnail,THUMB_MAX);
	}
	return jpeg_transform_files(argv[optind], outfile, transform,
				    comment, thumbnail, tsize, flags);
    } else {
	rc = 0;
	for (i = optind; i < argc; i++) {
	    fprintf(stderr,"processing %s\n",argv[i]);
	    if (flags & JFLAG_UPDATE_THUMBNAIL) {
		thumbnail = malloc(THUMB_MAX);
		tsize = create_thumbnail(argv[i],thumbnail,THUMB_MAX);
	    }
	    if (0 != jpeg_transform_inplace(argv[i], transform, comment,
					    thumbnail, tsize, flags))
		rc = 1;
	}
	return rc;
    }
}
